xdp/src/route_monitor.rs

```diff
@@ -0,0 +1,182 @@
+    /// Once drain_window expires, publishes the working routing table to the atomic router
+    /// If the netlink socket is invalid, recreates the socket and resyncs the atomic router
+    pub fn start(
+        atomic_router: Arc<ArcSwap<Router>>,
+        exit: Arc<AtomicBool>,
+        update_interval: Duration,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("solRouteMon".to_string())
+            .spawn(move || {
+                let mut state =
+                    RouteMonitorState::new(Router::new().expect("error creating Router"));
+
+                let timeout = Duration::from_millis(10);
+                while !exit.load(Ordering::Relaxed) {
+                    state.publish_if_needed(&atomic_router, update_interval);
+
+                    let mut pfd = pollfd {
+                        fd: state.sock.as_raw_fd(),
+                        events: POLLIN,
+                        revents: 0,
+                    };
+
+                    let ev = match poll(&mut pfd, timeout) {
+                        // timeout
+                        Ok(0) => continue,
+                        Ok(_) => pfd.revents,
+                        Err(e) => {
+                            error!("netlink poll error: {e}");
+                            state.reset(&atomic_router);
+                            continue;
+                        }
+                    };
+
+                    assert!(ev & POLLNVAL == 0);
+
+                    if (ev & (POLLHUP | POLLERR)) != 0 {
+                        error!(
+                            "netlink poll error (revents={}{})",
+                            if ev & POLLERR != 0 { "POLLERR" } else { "" },
+                            if ev & POLLHUP != 0 { "POLLHUP" } else { "" },
+                        );
+                        state.reset(&atomic_router);
+                        continue;
+                    }
+                    if (ev & POLLIN) == 0 {
+                        continue;
+                    }
+                    // Drain channel
+                    match state.sock.recv() {
+                        Ok(msgs) => {
+                            if !msgs.is_empty() {
+                                state.dirty |=
+                                    Self::process_netlink_updates(&mut state.router, &msgs);
+                            }
+                        }
+                        Err(e) => {
+                            error!("netlink recv error: {e}");
+                            state.reset(&atomic_router);
+                            continue;
+                        }
+                    }
+                }
+            })
+            .unwrap()
+    }
+
+    #[inline]
+    fn process_netlink_updates(router: &mut Router, msgs: &[NetlinkMessage]) -> bool {
+        let mut dirty = false;
+        for m in msgs {
+            match m.header.nlmsg_type {
+                RTM_NEWROUTE if is_supported_ipv4_route_header(m) => {
```

Comment:
in all these matches you should parse_ then call entry.is_valid() etc, then you
can remove the is_supported_bla funcs